### PR TITLE
Add support for declaring component metadata rules in settings

### PR DESCRIPTION
**Pull request by melix**
## Context

This pull request builds on top of #14854 .

In a similar fashion to repositories, settings may declare "global"
rules which are applied if, and only if, a project doesn't declare
any specific component metadata rule.

There are three different modes for rules:

- prefer project rules: this is the default, if a project declares
component metadata rules, then they are used instead of the settings
rules
- prefer settings rules: if set then any rule declared in a project
is going to be ignored and a warning is issued, explaining what is
declaring a rule
- fail on project rules: similar to the previous case but failing
instead of warning

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Review by jjohannes — APPROVED** [review]
This looks good and I did not find anything suspicious when trying the feature (except for a small thing in the Javadoc).

The DSL method names are consistent with the other names in the DSL, which is good.

File: subprojects/core-api/src/main/java/org/gradle/api/initialization/DependencyResolutionManagement.java [hunk 1]
```diff
@@ -56,4 +57,36 @@
...
+
+    /**
+     * Returns the shared component metadata handler
+     */
+    ComponentMetadataHandler getComponents();
+
+    /**
+     * If this method is called, any component metadata rule declared on a project
+     * will cause the project to use the rules declared by the project, ignoring
+     * those declared in settings. A warning will be issued in such a situation.
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/initialization/DependencyResolutionManagement.java:74]: It's a good catch, there are no warnings in this case, it's a leftover from a previous implementation.

**Pull request merged**